utils/s2n_safety.c

```diff
@@ -78,6 +78,49 @@ bool s2n_constant_time_equals(const uint8_t *a, const uint8_t *b, const uint32_t
     return (xor == 0);
 }
 
+/*  See specification of this function in s2n_safety.h */
+bool s2n_constant_time_equals_partial(const uint8_t *const a,
+                                      const uint8_t *const b,
+                                      const uint32_t len)
+{
+    bool arrays_equal = true;
+    /* iterate over each byte in the slices */
+    for (uint32_t i = 0; i < len; i++)
+    CONTRACT_ASSIGNS(i, arrays_equal)
+    CONTRACT_INVARIANT(i <= len)
+    CONTRACT_INVARIANT(arrays_equal ==
+                       __CPROVER_forall { unsigned j; (j >= 0 && j < i) ==> (a[j] == b[j]) })
```

Comment:
The "forall" quantifier is not evaluated at runtime at all - it's purely for proof, so there no extra overhead or risk of O(N^2) going on here.

Reply:
My code is actually slightly quicker than the original - 8 instructions in the inner loop instead of 10 from the original on AArch 64, generated by GCC 13.1 at -O2.

Reply:
TLS as a protocol has a ~ 2^16 byte limit on the size of records, but s2n accepts buffers from callers too; we just made it a 32-bit limit from day one and it's never really been an issue.

Reply:
Changed it to "uint32_t"